Distinct between `char`, `unsigned char` and `signed char`.

When compiling `unsigned char *str = "hi";` with `gcc -pedantic`, it complains:

```
test.c:8:24: warning: pointer targets in initialization differ in signedness [-Wpointer-sign]
   unsigned char *str = "hi";
```

According to this: http://stackoverflow.com/a/4337252 `char` is not the same as `singed char`. 

One solution is when `char` is explicitly used, to not transform it into `Int8()` internally, and thus as `signed char` when generated to C.

Alternatively, now everywhere `signed` is generated too, which is redundant for each integer type, except char, where it can be used to enforce `char` objects to carry a sign. Maybe use `uint8` -> `unsigned char`, `int8` -> `signed char`, `char` -> `char`.

--- a/basec/trans/BaseC/desugar/__tests__/types.spt
+++ b/basec/trans/BaseC/desugar/__tests__/types.spt
@@ -2,10 +2,10 @@ module types
 
 language MetaC
 
-test 8 bit int [[
+test char should stay char [[
   char x;
 ]] run strip-annos to Program(_, [
-    VarDeclaration([], Type([], Int8()), _, _)
+    VarDeclaration([], Type([], Char()), _, _)
   ])
 
 test signed 8 bit int [[

--- a/basec/trans/BaseC/desugar/types.str
+++ b/basec/trans/BaseC/desugar/types.str
@@ -17,17 +17,25 @@ rules
   is-type-qualifier =
     ?Const() <+ ?Volatile()
 
+  desugar-to-unsigned: Char()  -> UInt8()
   desugar-to-unsigned: Int8()  -> UInt8()
   desugar-to-unsigned: Int16() -> UInt16()
   desugar-to-unsigned: Int32() -> UInt32()
   desugar-to-unsigned: Int64() -> UInt64()
 
+  // when it is explicitly `signed`
+  desugar-to-signed: Char() -> Int8()
+
   desugar-unsigned =
     ?specifiers
     ; try(fetch(?Unsigned()) ; <map(try(desugar-to-unsigned))> specifiers)
+    ; try(fetch(?Signed())   ; <map(try(desugar-to-signed))> specifiers)
     ; filter(not(?Unsigned() <+ ?Signed()))
 
-  desugar-to-sized-ints: Char()  -> Int8()
+  // char is not the same as signed char
+  // https://github.com/arian/metac/issues/24
+  // desugar-to-sized-ints: Char()  -> Int8()
+
   desugar-to-sized-ints: Short() -> Int16()
   desugar-to-sized-ints: Int()   -> Int32()
   desugar-to-sized-ints: Long()  -> Int64()

--- a/basec/trans/BaseC/generate/__tests__/declarations.spt
+++ b/basec/trans/BaseC/generate/__tests__/declarations.spt
@@ -6,6 +6,20 @@ test variable declarations [[
   int32 a,b,c;
 ]] build generate-c to "signed int a; signed int b; signed int c;"
 
+test char variable should not get signed/unsigned as it's different [[
+  char a;
+]] build generate-c to "char a;"
+
+test char variable should keep signed specifier [[
+  signed char a;
+  int8 b;
+]] build generate-c to "signed char a; signed char b;"
+
+test char variable should keep unsigned specifier [[
+  unsigned char a;
+  uint8 b;
+]] build generate-c to "unsigned char a; unsigned char b;"
+
 test array variable without an array size [[
   int32 a[];
 ]] build generate-c to "signed int a[];"

--- a/basec/trans/BaseC/types/Expressions.ts
+++ b/basec/trans/BaseC/types/Expressions.ts
@@ -152,8 +152,16 @@ type functions
 
  // See section A6.1/A6.5 of "The C programming language"
  // TODO improve this
-  promote: (t1, t2) -> t
-  where
+  promote: (t1', t2') -> t
+    where (
+           t1' == Char() and Int8() => t1
+        or t1' => t1
+      )
+    and (
+           t2' == Char() and Int8() => t2
+        or t2' => t2
+      )
+    and
        ((t1 == Float64() or t2 == Float64()) and Float64() => t)
     or ((t1 == Float32() or t2 == Float32()) and Float32() => t)
     or ((t1 == UInt64()  or t2 == UInt64())  and UInt64() => t)

--- a/basec/trans/BaseC/types/Types.ts
+++ b/basec/trans/BaseC/types/Types.ts
@@ -22,6 +22,7 @@ relations
          t <is: Int()
       or t <is: Float()
 
+  Char() <is: Int()
   Int8() <is: Int()
   Int16() <is: Int()
   Int32() <is: Int()
@@ -42,6 +43,10 @@ relations
     where
        t1 == t2
     or t1 <widens-prim: t2
+
+  // Char should be treated equally to Int8()
+  Int8() <widens-prim: Char()
+  Char() <widens-prim: Int8()
 
   Int8() <widens-prim: Int16()
   Int8() <widens-prim: Int32()

--- a/basec/trans/BaseC/types/__tests__/expr-bin.spt
+++ b/basec/trans/BaseC/types/__tests__/expr-bin.spt
@@ -86,3 +86,17 @@ test using variables in expressions [[
 test subtract [[
   int32 f() { [[10 - 3]]; }
 ]] run get-type to UInt8()
+
+test promote int8 to int32 [[
+  void f(int8 a, int32 b) { [[a + b]]; }
+]] run get-type to Int32()
+
+test promote char to int16 [[
+  void f(char a, int16 b) { [[a + b]]; }
+]] run get-type to Int16()
+
+test promote uint8 to int16 [[
+  void f(uint8 a, int16 b) { [[a + b]]; }
+]] run get-type to Int16()
+
+

--- a/basec/trans/BaseC/types/__tests__/types.spt
+++ b/basec/trans/BaseC/types/__tests__/types.spt
@@ -62,3 +62,17 @@ test union type [[
   union X a;
   union X b = [[a]];
 ]] run get-type to Union(Identifier("X"{_}))
+
+test char type [[
+  char a;
+  char b = [[a]];
+]] run get-type to Char()
+
+test char type is assignable with int8 [[
+  char a;
+  int8 b = a;
+]] 0 errors
+
+test char variable can have a char [[
+  char a = 'a';
+]] 0 errors

--- a/basec/trans/BaseC/types/type-relations.str
+++ b/basec/trans/BaseC/types/type-relations.str
@@ -2,7 +2,7 @@ module BaseC/types/type-relations
 
 imports
 
-  runtime/task/tasks
+  runtime/task/create
 
 signature constructors
 

